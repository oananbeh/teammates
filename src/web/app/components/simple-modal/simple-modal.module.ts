import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NgbModule } from '@ng-bootstrap/ng-bootstrap';
import { SimpleModalComponent } from './simple-modal.component';
import { ProgressBarModule } from '../progress-bar/progress-bar.module';

/**
 * Module for the modal component
 */
@NgModule({
  declarations: [SimpleModalComponent],
  exports: [SimpleModalComponent],
  imports: [
    CommonModule,
    ProgressBarModule,
    NgbModule,
  ],
})

export class SimpleModalModule { }
