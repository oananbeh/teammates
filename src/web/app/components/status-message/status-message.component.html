<div class="alert alert-{{ statusMessage.color }}" *ngFor="let statusMessage of messages">
  <button class="btn-close" (click)="dismiss(statusMessage)"></button>
  <div style="padding-right: 30px;" [innerHtml]="statusMessage.message"></div>
</div>
