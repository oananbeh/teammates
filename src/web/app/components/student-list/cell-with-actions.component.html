<div class="text-center">
    <ng-template #actionButton let-isEnabled="isEnabled" let-tooltip="tooltip" let-name="name" let-id="id"
             let-tmRouterLink="tmRouterLink" let-queryParams="queryParams">
        <a class="btn btn-light btn-sm btn-margin-right custom-button" target="_blank" rel="noopener noreferrer"
            [id]="id"
            [ngClass]="{'disabled mouse-hover-only': !isEnabled}"
            [ngbTooltip]="isEnabled ? tooltip : 'You do not have the permissions to access this feature'"
            [tmRouterLink]="tmRouterLink"
            [queryParams]="queryParams">
        {{name}}
        </a>
    </ng-template>
    <ng-container *ngTemplateOutlet="actionButton; context: {
          id: 'btn-view-details-' + courseId + '-' + idx ,
          isEnabled: instructorPrivileges.canViewStudentInSections,
          tooltip: 'View the details of the student',
          name: 'View',
          tmRouterLink: '/web/instructor/courses/student/details',
          queryParams: {courseid: courseId, studentemail: email}
        }">
    </ng-container>
    <ng-container *ngTemplateOutlet="actionButton; context: {
          id: 'btn-edit-details-' + courseId + '-' + idx,
          isEnabled: instructorPrivileges.canModifyStudent,
          tooltip: 'Use this to edit the details of this student. To edit multiple students'
            + ' in one go, you can use the enroll page: '
            + 'Simply enroll students using the updated data and existing data will be updated accordingly',
          name: 'Edit',
          tmRouterLink: '/web/instructor/courses/student/edit',
          queryParams: {courseid: courseId, studentemail: email}
        }">
    </ng-container>
    <ng-container *ngIf="enableRemindButton">
        <ng-container>
            <button id="btn-send-invite-{{ courseId }}-{{ idx }}" class="btn btn-light btn-sm btn-margin-right custom-button"
                [ngClass]="{'disabled mouse-hover-only': !instructorPrivileges.canModifyStudent }"
                [disabled]=!isActionButtonsEnabled
                [ngbTooltip]="instructorPrivileges.canModifyStudent
                ? 'Email an invitation to the student requesting him/her to join the course using his/her'
                + ' Google Account. Note: Students can use TEAMMATES without \'joining\','
                + ' but a joined student can access extra features e.g. set up a user profile'
                : 'You do not have the permissions to access this feature'"
                (click)="remindStudentFromCourse()">
                Send Invite
            </button>
        </ng-container>
    </ng-container>
    <button id="btn-delete-{{ courseId }}-{{ idx }}" class="btn btn-light btn-sm btn-margin-right custom-button"
        [ngClass]="{'disabled mouse-hover-only': !instructorPrivileges.canModifyStudent }"
        [ngbTooltip]="instructorPrivileges.canModifyStudent
          ? 'Delete the student and the corresponding submissions from the course'
          : 'You do not have the permissions to access this feature'"
        [disabled]="!isActionButtonsEnabled"
        (click)="removeStudentFromCourse()">
        Delete
    </button>
    <ng-container *ngTemplateOutlet="actionButton; context: {
        id: 'btn-view-records-' + courseId + '-' + idx,
        isEnabled: instructorPrivileges.canModifyStudent,
        tooltip: 'View all data about this student',
        name: 'All Records',
        tmRouterLink: '/web/instructor/students/records',
        queryParams: {courseid: courseId, studentemail: email}
      }">
    </ng-container>
</div>
