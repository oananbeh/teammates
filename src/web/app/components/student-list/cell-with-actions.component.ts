import { CommonModule } from '@angular/common';
import { Component, Input } from '@angular/core';
import {
  NgbDropdownModule,
  NgbTooltipModule,
} from '@ng-bootstrap/ng-bootstrap';
import { InstructorPermissionSet } from 'src/web/types/api-request';
import { AjaxLoadingModule } from '../ajax-loading/ajax-loading.module';
import { TeammatesRouterModule } from '../teammates-router/teammates-router.module';

@Component({
  selector: 'tm-group-buttons',
  templateUrl: './cell-with-actions.component.html',
  standalone: true,
  imports: [
    CommonModule,
    TeammatesRouterModule,
    AjaxLoadingModule,
    NgbDropdownModule,
    NgbTooltipModule,
  ],
})

export class CellWithActionsComponent {
  @Input() idx: number = 0;
  @Input() courseId: string = '';
  @Input() email: string = '';
  @Input() isSendReminderLoading: boolean = false;
  @Input() enableRemindButton: boolean = false;
  @Input() isActionButtonsEnabled: boolean = true;

  @Input() instructorPrivileges: InstructorPermissionSet = {
    canModifyCourse: false,
    canModifyInstructor: false,
    canModifySession: false,
    canModifyStudent: false,
    canViewStudentInSections: false,
    canViewSessionInSections: false,
    canSubmitSessionInSections: false,
    canModifySessionCommentsInSections: false,
  };

  @Input() remindStudentFromCourse: () => void = () => {};
  @Input() removeStudentFromCourse : () => void = () => {};

}
