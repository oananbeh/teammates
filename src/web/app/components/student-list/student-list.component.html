<div>
  <tm-sortable-table
    [columns]="columnsData"
    [rows]="rowsData"
    [initialSortBy]="tableSortBy"
    [sortOrder]="tableSortOrder"
    [headerColorScheme]="headerColorScheme"
    [customHeaderStyle]="customHeaderStyle"
    (sortEvent)="sortStudentListEventHandler($event)">
  </tm-sortable-table>
</div>
