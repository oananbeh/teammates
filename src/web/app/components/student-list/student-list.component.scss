.align-center {
  text-align: center;
}

// applies only to mobile screens
@media (width <= 768px) {
  .custom-button {
    width: 90px;
    margin-bottom: 5px;
  }
}

.table .thead-gray th {
  color: #495057;
  background-color: #dadada;
}

.fa-sort {
  color: white;
}

/*
Allow only mouse hover events for the element.
!important is required to override Bootstrap .disabled class matches
which usually match 2 classes (eg .btn.disabled)
*/
.mouse-hover-only {
  pointer-events: auto !important;
}

.mouse-hover-only:active {
  pointer-events: none !important;
}

.sortable-header {
  cursor: pointer;

  i {
    margin-left: 0.3em;
  }

  .fa-stack {
    width: 20%;
  }

  button {
    background: none;
    color: inherit;
    border: none;
    padding: 0;
    text-align: left;
    width: 100%;
    font: inherit;
    cursor: inherit;
  }
}
