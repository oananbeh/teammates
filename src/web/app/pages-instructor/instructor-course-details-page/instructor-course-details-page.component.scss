.form-group {
  margin-bottom: 0.5rem;
}

.fill-plain {
  background-color: #eaeff5;
}

button {
  margin-right: 5px;
}

label {
  font-weight: bold;
}

.btn-link:hover {
  cursor: pointer;
}

.student-table-margin {
  margin-top: 50px;
}

@media (width <= 768px) {
  #btn-remind-all,
  #btn-download {
    margin-bottom: 0.5rem;
  }
}
