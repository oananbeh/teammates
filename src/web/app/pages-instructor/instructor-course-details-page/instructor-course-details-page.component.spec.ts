import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { of, throwError } from 'rxjs';
import SpyInstance = jest.SpyInstance;
import { InstructorCourseDetailsPageComponent } from './instructor-course-details-page.component';
import { InstructorCourseDetailsPageModule } from './instructor-course-details-page.module';
import { CourseStatistics } from '../../../services/course.service';
import { SimpleModalService } from '../../../services/simple-modal.service';
import { StatusMessageService } from '../../../services/status-message.service';
import { StudentService } from '../../../services/student.service';
import { createMockNgbModalRef } from '../../../test-helpers/mock-ngb-modal-ref';
import { Course, Instructor, InstructorPermissionRole, JoinState, Student } from '../../../types/api-output';
import { SimpleModalModule } from '../../components/simple-modal/simple-modal.module';
import { StudentListRowModel } from '../../components/student-list/student-list.component';
import { TeammatesCommonModule } from '../../components/teammates-common/teammates-common.module';

const course: Course = {
  courseId: 'CS101',
  courseName: 'Introduction to CS',
  timeZone: '',
  institute: 'Test Institute',
  creationTimestamp: 0,
  deletionTimestamp: 0,
};

const testStudent: Student = {
  name: 'Jamie',
  email: 'jamie@gmail.com',
  joinState: JoinState.NOT_JOINED,
  teamName: 'Team 1',
  sectionName: 'Tutorial Group 1',
  courseId: 'CS101',
};

const testInstructor: Instructor = {
  courseId: course.courseId,
  joinState: JoinState.JOINED,
  googleId: 'Hock',
  name: 'Hock',
  email: 'hock@gmail.com',
  role: InstructorPermissionRole.INSTRUCTOR_PERMISSION_ROLE_COOWNER,
  displayedToStudentsAs: 'Hock',
  isDisplayedToStudents: false,
};

describe('InstructorCourseDetailsPageComponent', () => {
  let component: InstructorCourseDetailsPageComponent;
  let fixture: ComponentFixture<InstructorCourseDetailsPageComponent>;
  let studentService: StudentService;
  let simpleModalService: SimpleModalService;
  let statusMessageService: StatusMessageService;

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      imports: [
        HttpClientTestingModule,
        TeammatesCommonModule,
        RouterTestingModule,
        InstructorCourseDetailsPageModule,
        SimpleModalModule,
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(InstructorCourseDetailsPageComponent);
    studentService = TestBed.inject(StudentService);
    simpleModalService = TestBed.inject(SimpleModalService);
    statusMessageService = TestBed.inject(StatusMessageService);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with default fields', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should snap with a course with one co-owner and no students, and populated course student list', () => {
    const stats: CourseStatistics = {
      numOfSections: 0,
      numOfTeams: 0,
      numOfStudents: 0,
    };
    const coOwner: Instructor = {
      courseId: course.courseId,
      joinState: JoinState.JOINED,
      googleId: 'Hodor',
      name: 'Hodor',
      email: 'hodor@gmail.com',
      role: InstructorPermissionRole.INSTRUCTOR_PERMISSION_ROLE_COOWNER,
      displayedToStudentsAs: 'Hodor',
      isDisplayedToStudents: true,
    };
    const courseDetails: any = {
      course,
      stats,
    };
    component.courseDetails = courseDetails;
    component.instructors = [coOwner];
    component.isLoadingCsv = false;
    component.isStudentsLoading = false;

    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap with a course with one co-owner and one student, and ajax failure', () => {
    const stats: CourseStatistics = {
      numOfSections: 1,
      numOfTeams: 1,
      numOfStudents: 1,
    };
    const coOwner: Instructor = {
      courseId: course.courseId,
      joinState: JoinState.JOINED,
      googleId: 'Bran',
      name: 'Bran',
      email: 'bran@gmail.com',
      role: InstructorPermissionRole.INSTRUCTOR_PERMISSION_ROLE_COOWNER,
      displayedToStudentsAs: 'Bran',
      isDisplayedToStudents: false,
    };
    const courseDetails: any = {
      course,
      stats,
    };
    const studentListRowModel: StudentListRowModel = {
      student: testStudent,
      isAllowedToViewStudentInSection: true,
      isAllowedToModifyStudent: true,
    };
    component.students = [studentListRowModel];
    component.courseDetails = courseDetails;
    component.instructors = [coOwner];
    component.isLoadingCsv = false;
    component.isStudentsLoading = false;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when students are still loading', () => {
    component.isStudentsLoading = true;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should display confirmation modal if delete all students is requested', () => {
    const stats: CourseStatistics = {
      numOfSections: 1,
      numOfTeams: 1,
      numOfStudents: 1,
    };
    const courseDetails: any = {
      course,
      stats,
    };
    const studentListRowModel: StudentListRowModel = {
      student: testStudent,
      isAllowedToViewStudentInSection: true,
      isAllowedToModifyStudent: true,
    };
    component.students = [studentListRowModel];
    component.courseDetails = courseDetails;
    component.instructors = [testInstructor];
    component.isLoadingCsv = false;
    component.isStudentsLoading = false;
    fixture.detectChanges();

    const promise: Promise<void> = Promise.resolve();

    const spySimpleModalService: SpyInstance = jest.spyOn(simpleModalService, 'openConfirmationModal')
        .mockReturnValue(createMockNgbModalRef({}, promise));

    const deleteAllButton: any = fixture.debugElement.nativeElement.querySelector('#btn-delete-all');
    deleteAllButton.click();
    fixture.detectChanges();

    expect(spySimpleModalService).toHaveBeenCalled();
  });

  it('should delete students in batches and show success message upon completion', () => {
    const stats: CourseStatistics = {
      numOfSections: 10,
      numOfTeams: 10,
      numOfStudents: 350,
    };
    const courseDetails: any = {
      course,
      stats,
    };
    component.courseDetails = courseDetails;
    fixture.detectChanges();

    const spyStudentService: SpyInstance = jest.spyOn(studentService, 'batchDeleteStudentsFromCourse')
        .mockReturnValue(of({ message: 'Successful' }));

    jest.spyOn(statusMessageService, 'showSuccessToast')
        .mockImplementation((args: string) => {
          expect(args).toEqual('All the students have been removed from the course');
        });

    jest.spyOn(simpleModalService, 'openLoadingModal')
        .mockReturnValue(createMockNgbModalRef());

    component.deleteAllStudentsFromCourse(course.courseId);

    // given a limit of 100 students per call and 350 students,
    // there should be four calls in total
    expect(spyStudentService).toHaveBeenCalledTimes(4);
  });

  it('should show error message when delete fails', () => {
    const stats: CourseStatistics = {
      numOfSections: 1,
      numOfTeams: 1,
      numOfStudents: 1,
    };
    const courseDetails: any = {
      course,
      stats,
    };
    component.courseDetails = courseDetails;
    fixture.detectChanges();

    jest.spyOn(studentService, 'batchDeleteStudentsFromCourse').mockReturnValue(throwError(() => ({
      error: {
        message: 'This is the error message.',
      },
    })));

    const spy: SpyInstance = jest.spyOn(statusMessageService, 'showErrorToast').mockImplementation((args: string) => {
      expect(args).toEqual('This is the error message.');
    });

    component.deleteAllStudentsFromCourse(course.courseId);

    expect(spy).toHaveBeenCalled();
  });
});
