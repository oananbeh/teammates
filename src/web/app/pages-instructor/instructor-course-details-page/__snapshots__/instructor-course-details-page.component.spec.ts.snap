// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstructorCourseDetailsPageComponent should snap when students are still loading 1`] = `
<tm-instructor-course-details-page
  courseDetails={[Function Object]}
  courseService={[Function CourseService]}
  hasLoadingStudentsFailed="false"
  instructorService={[Function InstructorService]}
  instructors={[Function Array]}
  isDeleting="false"
  isLoadingCsv="false"
  isStudentsLoading={[Function Boolean]}
  ngbModal={[Function _NgbModal]}
  progressBarService={[Function ProgressBarService]}
  route={[Function ActivatedRoute]}
  sectionsLoaded="0"
  simpleModalService={[Function SimpleModalService]}
  statusMessageService={[Function StatusMessageService]}
  studentService={[Function StudentService]}
  studentSortBy="0"
  studentSortOrder={[Function Number]}
  students={[Function Array]}
  tableComparatorService={[Function TableComparatorService]}
>
  <tm-loading-retry>
    <tm-loading-spinner>
      <div
        class="loading-container"
      >
        <div
          class="row spinner-border text-primary"
          role="status"
        />
        <div
          class="row text-secondary"
        >
          Loading...
        </div>
      </div>
    </tm-loading-spinner>
  </tm-loading-retry>
</tm-instructor-course-details-page>
`;

exports[`InstructorCourseDetailsPageComponent should snap with a course with one co-owner and no students, and populated course student list 1`] = `
<tm-instructor-course-details-page
  courseDetails={[Function Object]}
  courseService={[Function CourseService]}
  hasLoadingStudentsFailed="false"
  instructorService={[Function InstructorService]}
  instructors={[Function Array]}
  isDeleting="false"
  isLoadingCsv="false"
  isStudentsLoading="false"
  ngbModal={[Function _NgbModal]}
  progressBarService={[Function ProgressBarService]}
  route={[Function ActivatedRoute]}
  sectionsLoaded="0"
  simpleModalService={[Function SimpleModalService]}
  statusMessageService={[Function StatusMessageService]}
  studentService={[Function StudentService]}
  studentSortBy="0"
  studentSortOrder={[Function Number]}
  students={[Function Array]}
  tableComparatorService={[Function TableComparatorService]}
>
  <tm-loading-retry>
    <div>
      <h1>
        Course Details
      </h1>
      <div
        class="card"
      >
        <div
          class="card-body fill-plain"
        >
          <div
            class="row"
          >
            <label
              class="col-sm-2 text-sm-end"
            >
              Course ID:
            </label>
            <span
              class="col-sm-9 text-sm-start"
              id="course-id"
            >
               CS101 
            </span>
          </div>
          <div
            class="row"
          >
            <label
              class="col-sm-2 text-sm-end"
            >
              Course name:
            </label>
            <span
              class="col-sm-9 text-sm-start"
              id="course-name"
            >
               Introduction to CS 
            </span>
          </div>
          <div
            class="row"
          >
            <label
              class="col-sm-2 text-sm-end"
            >
              Course institute:
            </label>
            <span
              class="col-sm-9 text-sm-start"
              id="course-institute"
            >
               Test Institute 
            </span>
          </div>
          <div
            class="row"
          >
            <label
              class="col-sm-2 text-sm-end"
            >
              Sections:
            </label>
            <span
              class="col-sm-9 text-sm-start"
              id="num-sections"
            >
               0 
            </span>
          </div>
          <div
            class="row"
          >
            <label
              class="col-sm-2 text-sm-end"
            >
              Teams:
            </label>
            <span
              class="col-sm-9 text-sm-start"
              id="num-teams"
            >
               0 
            </span>
          </div>
          <div
            class="row"
          >
            <label
              class="col-sm-2 text-sm-end"
            >
              Total students:
            </label>
            <span
              class="col-sm-9 text-sm-start"
              id="num-students"
            >
               0 
            </span>
          </div>
          <div
            class="row"
          >
            <label
              class="col-sm-2 text-sm-end control-label"
            >
              Instructors:
            </label>
            <div
              class="col-sm-9 text-sm-start"
              id="instructors"
            >
              <div>
                 Co-owner: Hodor (hodor@gmail.com) 
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="alert alert-warning"
      >
        <i
          class="fa fa-exclamation-circle text-muted"
        />
         There are no students in this course. Click 
        <a
          href="/web/instructor/courses/enroll?courseid=CS101"
          tmrouterlink="/web/instructor/courses/enroll"
        >
          here
        </a>
         to enroll students. 
      </div>
    </div>
  </tm-loading-retry>
</tm-instructor-course-details-page>
`;

exports[`InstructorCourseDetailsPageComponent should snap with a course with one co-owner and one student, and ajax failure 1`] = `
<tm-instructor-course-details-page
  courseDetails={[Function Object]}
  courseService={[Function CourseService]}
  hasLoadingStudentsFailed="false"
  instructorService={[Function InstructorService]}
  instructors={[Function Array]}
  isDeleting="false"
  isLoadingCsv="false"
  isStudentsLoading="false"
  ngbModal={[Function _NgbModal]}
  progressBarService={[Function ProgressBarService]}
  route={[Function ActivatedRoute]}
  sectionsLoaded="0"
  simpleModalService={[Function SimpleModalService]}
  statusMessageService={[Function StatusMessageService]}
  studentService={[Function StudentService]}
  studentSortBy="0"
  studentSortOrder={[Function Number]}
  students={[Function Array]}
  tableComparatorService={[Function TableComparatorService]}
>
  <tm-loading-retry>
    <div>
      <h1>
        Course Details
      </h1>
      <div
        class="card"
      >
        <div
          class="card-body fill-plain"
        >
          <div
            class="row"
          >
            <label
              class="col-sm-2 text-sm-end"
            >
              Course ID:
            </label>
            <span
              class="col-sm-9 text-sm-start"
              id="course-id"
            >
               CS101 
            </span>
          </div>
          <div
            class="row"
          >
            <label
              class="col-sm-2 text-sm-end"
            >
              Course name:
            </label>
            <span
              class="col-sm-9 text-sm-start"
              id="course-name"
            >
               Introduction to CS 
            </span>
          </div>
          <div
            class="row"
          >
            <label
              class="col-sm-2 text-sm-end"
            >
              Course institute:
            </label>
            <span
              class="col-sm-9 text-sm-start"
              id="course-institute"
            >
               Test Institute 
            </span>
          </div>
          <div
            class="row"
          >
            <label
              class="col-sm-2 text-sm-end"
            >
              Sections:
            </label>
            <span
              class="col-sm-9 text-sm-start"
              id="num-sections"
            >
               1 
            </span>
          </div>
          <div
            class="row"
          >
            <label
              class="col-sm-2 text-sm-end"
            >
              Teams:
            </label>
            <span
              class="col-sm-9 text-sm-start"
              id="num-teams"
            >
               1 
            </span>
          </div>
          <div
            class="row"
          >
            <label
              class="col-sm-2 text-sm-end"
            >
              Total students:
            </label>
            <span
              class="col-sm-9 text-sm-start"
              id="num-students"
            >
               1 
            </span>
          </div>
          <div
            class="row"
          >
            <label
              class="col-sm-2 text-sm-end control-label"
            >
              Instructors:
            </label>
            <div
              class="col-sm-9 text-sm-start"
              id="instructors"
            >
              <div>
                 Co-owner: Bran (bran@gmail.com) 
              </div>
            </div>
          </div>
          <div
            style="margin-top: 10px;"
          >
            <div
              class="text-center"
            >
              <button
                class="btn btn-primary"
                id="btn-remind-all"
                type="button"
                value="Remind Students to Join"
              >
                <i
                  class="fa fa-envelope"
                />
                 Remind Students to Join 
              </button>
              <button
                class="btn btn-primary btn-margin-left"
                id="btn-download"
                type="button"
              >
                <i
                  class="fa fa-download"
                />
                 Download Student List 
              </button>
              <button
                class="btn btn-danger btn-margin-left"
                id="btn-delete-all"
                type="button"
                value="Delete All Students"
              >
                <i
                  class="fa fa-trash"
                />
                 Delete All Students 
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="card-body p-0 student-table-margin card border"
        id="student-list"
      >
        <tm-student-list>
          <div>
            <tm-sortable-table>
              <table
                class="table table-bordered table-striped"
              >
                <thead>
                  <tr
                    class="bg-light"
                  >
                    <th
                      aria-sort="none"
                      class="sort-by-section sortable-header"
                      style="text-align: start;"
                    >
                      <button
                        class="d-lg-flex align-items-center gap-lg-1 d-inline"
                      >
                        <span
                          container="body"
                        >
                          Section
                        </span>
                        <span
                          aria-hidden="true"
                          class="fa-stack"
                        >
                          <i
                            class="fas fa-sort"
                          />
                        </span>
                      </button>
                    </th>
                    <th
                      aria-sort="none"
                      class="sort-by-team sortable-header"
                      style="text-align: start;"
                    >
                      <button
                        class="d-lg-flex align-items-center gap-lg-1 d-inline"
                      >
                        <span
                          container="body"
                        >
                          Team
                        </span>
                        <span
                          aria-hidden="true"
                          class="fa-stack"
                        >
                          <i
                            class="fas fa-sort"
                          />
                        </span>
                      </button>
                    </th>
                    <th
                      aria-sort="none"
                      class="sort-by-name sortable-header"
                      style="text-align: start;"
                    >
                      <button
                        class="d-lg-flex align-items-center gap-lg-1 d-inline"
                      >
                        <span
                          container="body"
                        >
                          Student Name
                        </span>
                        <span
                          aria-hidden="true"
                          class="fa-stack"
                        >
                          <i
                            class="fas fa-sort"
                          />
                        </span>
                      </button>
                    </th>
                    <th
                      aria-sort="none"
                      class="sort-by-status sortable-header"
                      style="text-align: start;"
                    >
                      <button
                        class="d-lg-flex align-items-center gap-lg-1 d-inline"
                      >
                        <span
                          container="body"
                        >
                          Status
                        </span>
                        <span
                          aria-hidden="true"
                          class="fa-stack"
                        >
                          <i
                            class="fas fa-sort"
                          />
                        </span>
                      </button>
                    </th>
                    <th
                      aria-sort="none"
                      class="sort-by-email sortable-header"
                      style="text-align: start;"
                    >
                      <button
                        class="d-lg-flex align-items-center gap-lg-1 d-inline"
                      >
                        <span
                          container="body"
                        >
                          Email
                        </span>
                        <span
                          aria-hidden="true"
                          class="fa-stack"
                        >
                          <i
                            class="fas fa-sort"
                          />
                        </span>
                      </button>
                    </th>
                    <th
                      aria-sort="none"
                      style="text-align: center;"
                    >
                      <span
                        container="body"
                      >
                        Action(s)
                      </span>
                    </th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td>
                      <span>
                        Tutorial Group 1
                      </span>
                    </td>
                    <td>
                      <span>
                        Team 1
                      </span>
                    </td>
                    <td>
                      <span>
                        Jamie
                      </span>
                    </td>
                    <td>
                      Yet to Join
                    </td>
                    <td>
                      <span>
                        jamie@gmail.com
                      </span>
                    </td>
                    <td>
                      <ndc-dynamic />
                      <tm-group-buttons>
                        <div
                          class="text-center"
                        >
                          <a
                            class="btn btn-light btn-sm btn-margin-right custom-button"
                            href="/web/instructor/courses/student/details?courseid=CS101&studentemail=jamie@gmail.com"
                            id="btn-view-details-CS101-0"
                            rel="noopener noreferrer"
                            target="_blank"
                          >
                             View 
                          </a>
                          <a
                            class="btn btn-light btn-sm btn-margin-right custom-button"
                            href="/web/instructor/courses/student/edit?courseid=CS101&studentemail=jamie@gmail.com"
                            id="btn-edit-details-CS101-0"
                            rel="noopener noreferrer"
                            target="_blank"
                          >
                             Edit 
                          </a>
                          <button
                            class="btn btn-light btn-sm btn-margin-right custom-button"
                            id="btn-send-invite-CS101-0"
                          >
                             Send Invite 
                          </button>
                          <button
                            class="btn btn-light btn-sm btn-margin-right custom-button"
                            id="btn-delete-CS101-0"
                          >
                             Delete 
                          </button>
                          <a
                            class="btn btn-light btn-sm btn-margin-right custom-button"
                            href="/web/instructor/students/records?courseid=CS101&studentemail=jamie@gmail.com"
                            id="btn-view-records-CS101-0"
                            rel="noopener noreferrer"
                            target="_blank"
                          >
                             All Records 
                          </a>
                        </div>
                      </tm-group-buttons>
                    </td>
                  </tr>
                </tbody>
              </table>
            </tm-sortable-table>
          </div>
        </tm-student-list>
      </div>
    </div>
  </tm-loading-retry>
</tm-instructor-course-details-page>
`;

exports[`InstructorCourseDetailsPageComponent should snap with default fields 1`] = `
<tm-instructor-course-details-page
  courseDetails={[Function Object]}
  courseService={[Function CourseService]}
  hasLoadingStudentsFailed="false"
  instructorService={[Function InstructorService]}
  instructors={[Function Array]}
  isDeleting="false"
  isLoadingCsv="false"
  isStudentsLoading={[Function Boolean]}
  ngbModal={[Function _NgbModal]}
  progressBarService={[Function ProgressBarService]}
  route={[Function ActivatedRoute]}
  sectionsLoaded="0"
  simpleModalService={[Function SimpleModalService]}
  statusMessageService={[Function StatusMessageService]}
  studentService={[Function StudentService]}
  studentSortBy="0"
  studentSortOrder={[Function Number]}
  students={[Function Array]}
  tableComparatorService={[Function TableComparatorService]}
>
  <tm-loading-retry>
    <tm-loading-spinner>
      <div
        class="loading-container"
      >
        <div
          class="row spinner-border text-primary"
          role="status"
        />
        <div
          class="row text-secondary"
        >
          Loading...
        </div>
      </div>
    </tm-loading-spinner>
  </tm-loading-retry>
</tm-instructor-course-details-page>
`;
